### PR TITLE
Turn InstanceContainer, DefinitionContainer and ContainerStack into QObject subclasses

### PR DESCRIPTION
**Pull request by awhiemstra**
This is the first step in the CuraContainerStack work. It makes all three main container classes QObject subclasses that can more easily be exposed to QML.

### CODE REVIEW

**Comment by nallath** [comment]
It needs a bit of work; 
2017-03-17 11:18:15,192 - CRITICAL - cura.CrashHandler.show [36]: An uncaught exception has occurred!
2017-03-17 11:18:15,194 - CRITICAL - cura.CrashHandler.show [39]: Traceback (most recent call last):
2017-03-17 11:18:15,194 - CRITICAL - cura.CrashHandler.show [39]:   File "cura_app.py", line 70, in <module>
2017-03-17 11:18:15,195 - CRITICAL - cura.CrashHandler.show [39]:     app = cura.CuraApplication.CuraApplication.getInstance()
2017-03-17 11:18:15,195 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Uranium/UM/Application.py", line 325, in getInstance
2017-03-17 11:18:15,195 - CRITICAL - cura.CrashHandler.show [39]:     Application._instance = cls()
2017-03-17 11:18:15,196 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Cura/cura/CuraApplication.py", line 219, in __init__
2017-03-17 11:18:15,196 - CRITICAL - cura.CrashHandler.show [39]:     empty_variant_container.addMetaDataEntry("type", "variant")
2017-03-17 11:18:15,196 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Uranium/UM/Settings/InstanceContainer.py", line 200, in addMetaDataEntry
2017-03-17 11:18:15,197 - CRITICAL - cura.CrashHandler.show [39]:     self.metaDataChanged.emit()
2017-03-17 11:18:15,197 - CRITICAL - cura.CrashHandler.show [39]: RuntimeError: super-class __init__() of type _EmptyInstanceContainer was never called

**Comment by awhiemstra** [comment]
Fixed. I was assuming fixing pickle support for InstanceContainer would also have fixed deepcopy of InstanceContainer-subclasses but apparently not. Pickle is some weird black magic every now and then...

**Comment by nallath** [comment]
Still nope :( 
2017-03-20 11:39:38,696 - CRITICAL - cura.CrashHandler.show [36]: An uncaught exception has occurred!
2017-03-20 11:39:38,696 - CRITICAL - cura.CrashHandler.show [39]: Traceback (most recent call last):
2017-03-20 11:39:38,697 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Uranium/UM/Settings/Models/ContainerStacksModel.py", line 26, in __init__
2017-03-20 11:39:38,697 - CRITICAL - cura.CrashHandler.show [39]:     self._update()
2017-03-20 11:39:38,697 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Uranium/UM/Settings/Models/ContainerStacksModel.py", line 44, in _update
2017-03-20 11:39:38,697 - CRITICAL - cura.CrashHandler.show [39]:     container.nameChanged.disconnect(self._onContainerNameChanged)
2017-03-20 11:39:38,698 - CRITICAL - cura.CrashHandler.show [39]: TypeError: 'method' object is not connected
2017-03-20 11:39:38,870 - CRITICAL - cura.CrashHandler.show [36]: An uncaught exception has occurred!
2017-03-20 11:39:38,870 - CRITICAL - cura.CrashHandler.show [39]: Traceback (most recent call last):
2017-03-20 11:39:38,871 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Uranium/UM/Settings/Models/ContainerStacksModel.py", line 26, in __init__
2017-03-20 11:39:38,871 - CRITICAL - cura.CrashHandler.show [39]:     self._update()
2017-03-20 11:39:38,871 - CRITICAL - cura.CrashHandler.show [39]:   File "/home/jaime/Development/Uranium/UM/Settings/Models/ContainerStacksModel.py", line 44, in _update
2017-03-20 11:39:38,872 - CRITICAL - cura.CrashHandler.show [39]:     container.nameChanged.disconnect(self._onContainerNameChanged)
2017-03-20 11:39:38,872 - CRITICAL - cura.CrashHandler.show [39]: TypeError: 'method' object is not connected

File: UM/Settings/Models/ContainerStacksModel.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def __init__(self, parent = None):
         self.addRoleName(self.NameRole, "name")
         self.addRoleName(self.IdRole, "id")
         self.addRoleName(self.MetaDataRole, "metadata")
-        self._container_stacks = ContainerRegistry.getInstance().findContainerStacks()
+        self._container_stacks = []]
```
Comment (fieldOfView) [inline UM/Settings/Models/ContainerStacksModel.py]: That can't be right...

Comment (awhiemstra) [inline UM/Settings/Models/ContainerStacksModel.py]: No it's not, which is why I pushed a fix immediately after. :p

File: UM/Decorators.py [hunk 2]
```diff
@@ -54,6 +54,23 @@ def call_direct(decorated_function):
...
+#
+#   \param cls The class this function overrides a function from.
+def override(cls):
+    def override_decorator(function):
+        function_signature = inspect.signature(function)
+        if not inspect.getmembers(cls, lambda i: inspect.isfunction(i) and sameSignature(inspect.signature(i), function_signature)):
+            raise InvalidOverrideError("Method {method} is marked as override but was not found in base class {cls}".format(method = function.__qualname__, cls = cls.__qualname__))
+        return function
+    return override_decorator
+
```
Comment (sedwards2009) [inline UM/Decorators.py]: Can't mypy check this kind of thing statically?

**Pull request merged**